scripts/release/other_identifiers/retrieve_indirectIdentifiers_from_mart.pl

```diff
@@ -1,77 +1,156 @@
-#!/usr/local/bin/perl -w
+#!/usr/bin/perl
 use strict;
+use warnings;
 
 use lib "/usr/local/reactomes/Reactome/development/GKB/modules";
 
 use GKB::Config;
-use GKB::DBAdaptor;
-use GKB::EnsEMBLMartUtils qw/:all/;
-use GKB::Utils;
+use GKB::Config_Species;
+use GKB::EnsEMBLMartUtils qw/get_identifiers get_species_mart_name get_wget_query_for_identifier/;
 
-use Data::Dumper;
+use Capture::Tiny qw/capture/;
 use Getopt::Long;
+use List::MoreUtils qw/any/;
 use Try::Tiny;
 
 use Log::Log4perl qw/get_logger/;
-Log::Log4perl->init(\$LOG_CONF);
-my $logger = get_logger(__PACKAGE__);
-
-our($opt_user,$opt_host,$opt_pass,$opt_port,$opt_db,$opt_debug,$opt_sp);
-
-(@ARGV) || die "Usage: $0 -sp 'species name'
--user db_user -host db_host -pass db_pass -port db_port -db db_name\n";
-
-&GetOptions("user:s", "host:s", "pass:s", "port:i", "db:s", "debug", "sp=s");
-
-$opt_db || die "Need database name (-db).\n";    
-
-# Get connection to reactome db
-my $dba = GKB::DBAdaptor->new
-    (
-     -user   => $opt_user,
-     -host   => $opt_host,
-     -pass   => $opt_pass,
-     -port   => $opt_port,
-     -dbname => $opt_db,
-#     -DEBUG => $opt_debug
-     );
-
-# Fetch the species
-$opt_sp ||= 'Homo sapiens';
-
-my $sp = $dba->fetch_instance_by_attribute('Species',[['name',[$opt_sp]]])->[0]
-    || $logger->error_die("No species '$opt_sp' found.\n");
-my $sp_mart_name;
-if ($sp->displayName =~ /^(\w)\w+ (\w+)$/) {
-    $sp_mart_name = lc("$1$2");
-} else {
-    $logger->error_die("Can't form species abbreviation for mart from '" . $sp->displayName . "'.\n");
+
+run(@ARGV) unless caller();

```

Comment:
This is a fairly perlistic thing going on -- I think a comment explaining it would be helpful

Reply:
Fair enough :).  I've added this explanation.  Let me know if it needs tweaking.
```
# Run the main program if the script is invoked by the Perl intepreter
# This allows the program to be run normally and included in another
# Perl file for unit testing its functions using the 'require' syntax
run(@ARGV) unless caller();
```

Reply:
That... doesn't explain much. What is `caller()`? And the comment is written "normally" (using if-...) but the code uses inverted logic (... unless ...) so...

Does this statement mean "if the currently running program was called invoked by the Perl interpreter, execute the command line arguments as a system call"...?

Reply:
Sorry - I got the details a bit wrong.  

Really, this is just a switch to allow the script to run as a typical Perl script does or to allow it to just be used as a module with functions available.  This is a common pattern in Perl for when you want to be able to make a script's functions externally visible for unit testing in a '.t' file that's typical of Perl testing.

The caller function will return a value if a separate Perl script or package has loaded it.  So when it is invoked by the Perl interpreter, it returns nothing and the 'run' function gets executed.  The run is just like a stand-in for 'main' so the script's top-level is wrapped in a function.  When it is loaded by another Perl script that I use for testing, the caller returns a value and the run function does NOT execute but I have access to the functions for unit testing.  I'll change the `unless caller()` to `if !caller()` so we don't have to do the mental inversion. 

This page explains it, but the relevant part is below the link:

http://www.drdobbs.com/scripts-as-modules/184416165

The script-or-module magic checks the result of the caller() Perl built-in function. If something else in the Perl script calls the file, the caller() function (in scalar context) returns the calling package name. That's true if another Perl file loads this one with use() or require(). If I run my modulino as a script, there is no other file loading it, so caller() returns undef. If caller() returns a false value, then I execute the __PACKAGE__->run() method (the Perl compiler replaces the __PACKAGE__ token with the current package name).